### ISSUE TITLE
pull: pull all parameterized stages (`--group`)

### ISSUE THREAD

**Issue by sjawhar** [iterative/dvc#7462]
We use parametrized stages in our repos a _lot_ (`foreach`/`do`). It's a hassle to have to `dvc pull stage@parameter` for each parameter. It would be great if we could add a `--group` flag to `fetch`, `pull`, and `checkout` that would run for all parametrized versions of that stage. I thought `--glob` would do this, but after digging into the code I can see that's not what we want.

I called it `--glob` by taking the name of the kwarg from [`StageLoad.load_all()`](https://github.com/iterative/dvc/blob/3fc8dcfc59e672df082c5efc4c19550f511b07bf/dvc/repo/stage.py#L257), which is the method called under the hood by these various commands: `dvc pull` > `Repo.pull` > `Repo.fetch` > `Repo.used_objs` > `Index.used_objs` > `StageLoad.collect_granular` > `stage._collect_specific_target` > `stage._maybe_collect_from_dvc_yaml` > `StageLoad.load_all()`. :sweat_smile: 

If there are no objections or if someone isn't already working on it, I'm happy to contribute. I'd just pass the optional `accept_group: bool = False` down through that call chain.

**Comment by karajan1001** [comment]
For the target of `dvc pull` it can accept output file as argument. And if there were already these files in your workspace ( you want to update instead of newly downloaded)  you can use a wild match pattern to match them.

**Comment by sjawhar** [comment]
Yes, but the problem there is just as you stated: _**IF**_ you already have them in the workspace.

**Comment by dberenbaum** [comment]
> I thought `--glob` would do this, but after digging into the code I can see that's not what we want.

Seems like I should know this 😅 , but would someone mind explaining why `--glob` cannot do this?

**Comment by sjawhar** [comment]
> Seems like I should know this sweat_smile , but would someone mind explaining why `--glob` cannot do this?

`--glob` was implemented in #5032 to address #4816. It's about paths, which can be a bit confusing because `StageLoad.load_all()` also takes an optional `glob` argument, which does indeed glob match the stage names. :smile_cat:

**Comment by karajan1001** [comment]
> > I thought `--glob` would do this, but after digging into the code I can see that's not what we want.
> 
> Seems like I should know this 😅 , but would someone mind explaining why `--glob` cannot do this?

For glob, it only matches the targets.
```bash
--glob               Allows targets containing shell-style wildcards.
```
And besides this a big limitation is mentioned before: `IF you already have them in the workspace.`

For example for a stage `train` output `model.pkl`
We have the following result.
```bash
$ dvc repro --glob mode?.* # can match `model.pkl` 
Data and pipelines are up to date.
$ dvc repro --glob mode??.* # two ?? here can not match `model.pkl`
zsh: no matches found: mode??.*
$ dvc repro --glob trai* # cannot match stage name `train`
zsh: no matches found: trai*
$ rm model.pkl # if we delete the file `model.pkl`
$ dvc repro --glob mode?.* # cannot match it for now.
zsh: no matches found: mode?.*
```

**Comment by dberenbaum** [comment]
Hm, I was getting confused because of behavior without wildcards:

```
$ dvc repro --glob mode
Data and pipelines are up to date.
$ dvc repro --glob trai
Data and pipelines are up to date.
$ rm model.pkl
$ dvc repro --glob mode
Data and pipelines are up to date.
$ dvc repro --glob mode*
zsh: no matches found: mode*
```

It seems without wildcards that `--glob` is matching everything. Is this expected?

**Comment by skshetry** [comment]
@dberenbaum, the last one is coming from your shell, you have to quote it to avoid wildcard expansion by the shell.

**Comment by dberenbaum** [comment]
Isn't the same true for the example from @karajan1001?

**Comment by karajan1001** [comment]
> @dberenbaum, the last one is coming from your shell, you have to single-quote it to avoid wildcard expansion by the shell.

I think so, we need to add a `"` to prevent it from being replaced by the shell. 
```bash
$ dvc repro --glob "trai*"
'data/data.xml.dvc' didn't change, skipping
Stage 'prepare' didn't change, skipping
Stage 'featurize' didn't change, skipping
Stage 'train' is cached - skipping run, checking out outputs
Use `dvc push` to send your updates to remote storage.
$ dvc repro "trai*"
ERROR: Stage 'trai*' not found inside 'dvc.yaml' file
```
So, @sjawhar . you can try this method `dvc pull stage@1_*`, I tried it on my computer and get the correct file downloaded.
```bash
$ rm raw*.cln
dvc pull --glob "raw*.cln"
A       raw1.cln
A       raw2.cln
2 files added
```

**Comment by dberenbaum** [comment]
@karajan1001 I see that `repro --glob` works as expected, but `pull --glob` pulls more than expected unless I'm misunderstanding:

```
$ dvc pull --glob 'feat*'
A       model.pkl
A       data/prepared/
A       data/data.xml
A       data/features/
```

**Comment by sjawhar** [comment]
@karajan1001 that only _appears_ to work for you because it's in fact pulling _all_ stages. You can see this by setting a breakpoint here: https://github.com/iterative/dvc/blob/079714772cd471c3e3cea06912b7308b9eaf62cb/dvc/repo/pull.py#L32

No targets are matches, so `expanded_targets` is an empty list and everything gets pulled as a result:
![image](https://user-images.githubusercontent.com/5074378/158824362-9b566058-d6a3-4e69-89d9-d37494c82bb4.png)

**Comment by sjawhar** [comment]
@dberenbaum  `repro` uses `StageLoad.collect()`, which uses `StageLoad.load_all()`. So it is indeed doing the type of globbing you (and I) would expect. `pull` instead calls that other `glob_targets` function, which globs on path. I'm not sure why a `--glob` option for paths is needed when the shell can do it for you?

**Comment by pared** [comment]
Related #7323

**Comment by dberenbaum** [comment]
@sjawhar We discussed this one, and you managed to find multiple issues with the globbing features, which need a lot of work 😅 .

Addressing a group should be handled automatically as part of #7323, so simply `dvc pull stage` should work. Please keep an eye there for changes to address your initial request 🙏 .

Globbing outputs generally should not depend on what is checked out in the workspace. Otherwise, I agree it's not that useful for commands other than `dvc add`.

There's also confusion because `--glob` works differently in `repro` from `pull` and other commands, as you noticed. There was discussion in #5723 of separating into `--glob-stages` and/or `--glob-outs`, but ultimately it may be better to have globbing on by default for both.

**Comment by sjawhar** [comment]
OK, so how about I start with a PR that addresses that checklist you posted in #7323? That would actually take care of my needs. Maybe this issue can be closed in favor of #7323 and #7491.

**Comment by dberenbaum** [comment]
@sjawhar I'm not sure yet what the plan is to address #7323, and it might require some higher-level changes. @skshetry Can you comment on whether there is anything @sjawhar can do to help?
